# Assignment 1

K. Sai Somanath, 18MCMT28

In [1]:
import numpy as np

### Generating random samples from a normal distribution

In [2]:
def genNormalSamples(number_of_samples, mean, cov):
    return np.random.multivariate_normal(mean, cov, number_of_samples).T

### The Euclidean distance between two points

In [3]:
def euclidean_dist(x, y):
    sub = x - y
    return numpy.sqrt(numpy.einsum('ij,ij->i', sub, sub))

### The Mahalanobis Distance

In [4]:
def inverse(a):
    if not a.shape:
        return a
    else:
        return np.linalg.inv(a)

In [5]:
def norm(temp, cov_inv):
    if not temp.shape:
        # print('here')
        return np.sqrt(temp ** 2 * cov_inv)
    else:
        # print('Shape:', np.dot(np.dot(temp.T,cov_inv),temp).shape)
        return np.sqrt(np.dot(np.dot(temp.T,cov_inv),temp))

In [6]:
def mahalanobis_dist(x, mu, cov):
    cov_inv = inverse(cov)
    temp = x - mu
    # print(norm(temp, cov_inv))
    return norm(temp, cov_inv)

### The Discriminent function

$
g_i(x) = -\frac{1}{2}(x_i - \mu_i)^t\Sigma^{-1}(x_i - \mu_i) - \frac{d}{2}\ln{2\pi} - \frac{1}{2}\ln{\lvert \Sigma_i\rvert} + \ln{P(\omega_i)}
$

In [7]:
def det(mat):
    if not mat.shape:
        return mat
    else:
        return np.linalg.det(mat)

In [8]:
def disc_function(x, mu, cov, d, p_w):
    mahala_dist = mahalanobis_dist(x, mu, cov)
    return -((0.5) * mahala_dist) - ((d / 2) * np.log(np.pi * 2)) - (0.5 * np.log(det(cov))) + np.log(p_w)

In [9]:
data = np.genfromtxt('data_dhs_chap2.csv', delimiter=',', skip_header=1)

In [10]:
data.shape

(30, 4)

## Dichotomizer - Univariate

In [11]:
# Get class data
x_samples = data[:5, :1]
y_samples = data[10:15, :1]
samples = np.concatenate((x_samples, y_samples))

# Generate traget values
targets = [0] * 5 + [1] * 5
predicted = []

# Mean and covariace
mu1 = np.mean(x_samples, axis=0)
cov1 = np.cov(x_samples.T)
mu2 = np.mean(y_samples, axis=0)
cov2 = np.cov(y_samples.T)

for sample in samples:    
    g1 = disc_function(sample, mu1, cov1, 1, 0.5)
    g2 = disc_function(sample, mu2, cov2, 1, 0.5)
    if g1 - g2 > 0:
        predicted.append(0)
    elif g1 - g2 < 0:
        predicted.append(1)
    else:
        predicted.append(-1)
    
predicted

[0, 0, 0, 0, 0, 1, 0, 0, 0, 0]

### Error rate

The error rate turns out to be 40%

In [12]:
((np.array(predicted) != np.array(targets)).mean()) * 100

40.0

## Dichotomizer - Two features

In [13]:
# Get class data
x_samples = data[:5, :2]
y_samples = data[10:15, :2]
samples = np.concatenate((x_samples, y_samples))

# Generate traget values
targets = [0] * 5 + [1] * 5
predicted = []

# Mean and covariace
mu1 = np.mean(x_samples, axis=0)
cov1 = np.cov(x_samples.T)
mu2 = np.mean(y_samples, axis=0)
cov2 = np.cov(y_samples.T)

for sample in samples:
    g1 = disc_function(sample.reshape(2,1), mu1.reshape(2,1), cov1, 2, 0.5)
    g2 = disc_function(sample.reshape(2,1), mu2.reshape(2,1), cov2, 2, 0.5)
    if g1 - g2 > 0:
        predicted.append(0)
    elif g1 - g2 < 0:
        predicted.append(1)
    else:
        predicted.append(-1)
        
predicted

[0, 0, 0, 0, 1, 1, 0, 0, 0, 1]

### Error rate
The error rate turns out to be 40%

In [14]:
(np.array(predicted) != np.array(targets)).mean() * 100

40.0

## Dichotomizer - Three features

In [15]:
# Get class data
x_samples = data[:5, :3]
y_samples = data[10:15, :3]
samples = np.concatenate((x_samples, y_samples))

# Generate traget values
targets = [0] * 5 + [1] * 5
predicted = []

# Mean and covariace
mu1 = np.mean(x_samples, axis=0)
cov1 = np.cov(x_samples.T)
mu2 = np.mean(y_samples, axis=0)
cov2 = np.cov(y_samples.T)

for sample in samples:
    g1 = disc_function(sample.reshape(3,1), mu1.reshape(3,1), cov1, 3, 0.5)
    g2 = disc_function(sample.reshape(3,1), mu2.reshape(3,1), cov2, 3, 0.5)
    if g1 - g2 > 0:
        predicted.append(0)
    elif g1 - g2 < 0:
        predicted.append(1)
    else:
        predicted.append(-1)
        
predicted

[0, 0, 0, 1, 0, 1, 1, 1, 1, 1]

### Error rate
The error rate turns out to be 10%

In [16]:
(np.array(predicted) != np.array(targets)).mean() * 100

10.0

## Classifying new points into categories with uniform probabilities

In [17]:
# Get class data
x_samples = data[:10, :3]
y_samples = data[10:20, :3]
z_samples = data[20:, :3]

# Mean and covariace
mu1 = np.mean(x_samples, axis=0)
cov1 = np.cov(x_samples.T)
mu2 = np.mean(y_samples, axis=0)
cov2 = np.cov(y_samples.T)
mu3 = np.mean(z_samples, axis=0)
cov3 = np.cov(z_samples.T)

## The DFs
g1 = disc_function(sample.reshape(3,1), mu1.reshape(3,1), cov1, 2, 0.33)
g2 = disc_function(sample.reshape(3,1), mu2.reshape(3,1), cov2, 2, 0.33)
g3 = disc_function(sample.reshape(3,1), mu3.reshape(3,1), cov3, 2, 0.33)

data = [[1, 2, 1], [5, 3, 2], [0, 0, 0], [1, 0, 0]]
data = np.array(data)

In [18]:
distances = []
for point in data:
    for mu, cov in zip([mu1, mu2, mu3], [cov1, cov2, cov3]):
        distances.append(np.asscalar(mahalanobis_dist(point.reshape(3,1), mu.reshape(3,1), cov)))
distances = np.array(distances).reshape(4, 3)
np.argmin(distances, axis=1)

array([1, 2, 1, 1])

### Classification

We can see that points are classified as follows:

(1, 2, 1) --> 1

(5, 3, 2) --> 2

(0, 0, 0) --> 1

(1, 0, 0) --> 1

In [19]:
distances = []
for point in data:
    for mu, cov in zip([mu1, mu2, mu3], [cov1, cov2, cov3]):
        distances.append(np.asscalar(mahalanobis_dist(point.reshape(3,1), mu.reshape(3,1), cov)))
distances = np.array(distances).reshape(4, 3)
np.argmin(distances, axis=1)

array([1, 2, 1, 1])

## Conclusion

The classification remains the same as the *Mahalanobis* distances is independent of the **Prior probabilites**

## Iris Dataset

In [20]:
# Laod data
iris = np.genfromtxt('iris.csv', delimiter=',', skip_header=1, usecols=(0,1,2,3))
iris = iris.reshape(3, 50, 4)

In [21]:
means = np.mean(iris, axis=1)
means

array([[5.006, 3.418, 1.464, 0.244],
       [5.936, 2.77 , 4.26 , 1.326],
       [6.588, 2.974, 5.552, 2.026]])

In [22]:
# np.cov(iris.T).shape
covs = [np.cov(x.T) for x in iris]
covs = np.array(covs)
covs.shape

(3, 4, 4)

## Case 1:

The LDF for Case 1 is given by the equation: 

$g_i(x) = W_i^Tx+W_{i0}$

where $W_i= \frac{\mu_i}{\sigma^2} \implies W_i^T= \frac{\mu_i^T}{\sigma^2}$

and $W_{i0}= \frac{-\mu_i^T\mu_i}{2\sigma^2}+ ln(P(\omega_i))$


In [23]:
var = np.mean([covs[i].diagonal() for i in range(covs.shape[0])])

### Calculating $W^T$s

In [24]:
for mu in means:
    print(mu / var)

[32.93023131 22.48412517  9.63041523  1.6050692 ]
[39.04791311 18.22148237 28.02292956  8.72263019]
[43.33686853 19.56342547 36.52190256 13.32733692]


### Calculating $W_0$s

In [25]:
for mu in means:
    val = ((- 0.5 * (np.dot(mu, mu.T))) / var) + np.log(1/3)
    print(val)

-129.19363356697767
-207.70151526763146
-287.82646472325564


From the above calculations it is evident that we will have the following equations:

$g_1(x) = 32.93* x_1 + 22.48 * x_2 + 9.63 * x_3 + 1.60 *x_4 - 129.19$

$g_2(x) = 39.05* x_1 + 18.22 * x_2 + 28.02 * x_3 + 8.72 *x_4 - 207.70$

$g_3(x) = 43.34* x_1 + 19.56 * x_2 + 36.52 * x_3 + 13.33 *x_4 - 287.83$

We will be able to construct the equations in a **similar manner for all other cases**.

Below, we list the values of the $W$ and the $W_{i0}$ matrices, using which the equations are constructed for **case 2**.



## Case 2:

$g_i(x) = W_i^Tx+W_{i0}$ 

where $W_i= \Sigma^{-1}\mu_i \implies  W_i^T= \mu_i^T(\Sigma^{-1})^T$

and $W_{i0}= \frac{-1}{2}\mu_i^T\Sigma^{-1}\mu_i+ ln(P(\omega_i))$

In [26]:
cov_mean = np.mean(covs, axis=0)

### Calculating $W^T$s

In [27]:
for mu in zip(means):
    print(np.dot(mu, np.linalg.inv(cov_mean).T))

[[ 23.46638411  23.56828007 -16.20296668 -18.02533894]]
[[15.70349917  6.95425598  5.28429325  6.29830031]]
[[12.49061995  3.44398145 12.82207622 21.06272174]]


## Calculating $W_{0}$

In [28]:
for mu in means:
    dist = -0.5 * np.dot(np.dot(mu, np.linalg.inv(cov_mean)), mu.T)
    print(dist + np.log(1 / 3))

-86.05349938690331
-72.76956006760176
-104.29453550375828


## Case 3:

$
g_i(x) = x^tW_ix + w_i^tx + w_{i0}
$
Where,

$
W_i = -\frac{1}{2}\Sigma_i
$

$
w_i = \Sigma_i^{-1}\mu_i
$

$
w_{i0} = -\frac{1}{2}\mu^t_i\Sigma_i^{-1}\mu_i  -\frac{1}{2}\ln \lvert \Sigma_i \rvert + \ln(P(\omega_i))
$

### Calculating $W$s

In [29]:
Wi = - 0.5 * covs
Wi

array([[[-0.06212449, -0.05014898, -0.00806939, -0.00527347],
        [-0.05014898, -0.0725898 , -0.00584082, -0.00571837],
        [-0.00806939, -0.00584082, -0.01505306, -0.00284898],
        [-0.00527347, -0.00571837, -0.00284898, -0.00574694]],

       [[-0.13321633, -0.04259184, -0.09144898, -0.0278898 ],
        [-0.04259184, -0.04923469, -0.04132653, -0.02060204],
        [-0.09144898, -0.04132653, -0.11040816, -0.03655102],
        [-0.0278898 , -0.02060204, -0.03655102, -0.01955306]],

       [[-0.20217143, -0.04688163, -0.1516449 , -0.02454694],
        [-0.04688163, -0.05200204, -0.0356898 , -0.02381429],
        [-0.1516449 , -0.0356898 , -0.15229388, -0.02441224],
        [-0.02454694, -0.02381429, -0.02441224, -0.03771633]]])

### Calculating $w$s

In [30]:
wi = [np.dot(np.linalg.inv(c), mu.T) for c, mu in zip(covs, means)]
wi

[array([ 44.6351704 ,  -7.71338075,  33.07861577, -28.45246274]),
 array([ 18.0128645 ,  15.96070005,   3.26878502, -14.71255747]),
 array([ 7.37247478, 13.2452613 ,  6.23406948,  9.66197608])]

### Calculating $w_0$s

In [31]:
wi0 = []
for c, mu in zip(covs, means):
    dist = -0.5 * np.dot(np.dot(mu, np.linalg.inv(c)), mu.T)
    wi0.append(dist - (0.5 * np.log(np.linalg.det(c))) + np.log(1 / 3))
wi0 = np.array(wi0)
wi0

array([-113.86317185,  -68.43728769,  -67.7090772 ])

Now we have the following values:
1. $W_i, W_j, W_k$
2. $w_i, w_j, w_k$
3. $w_{i0}, w_{j0}, w_{k0}$
Using the above values we get the Quadractic discriminent function $g_i(x), g_j(x), g_k(x)$

Hence, we now have constructed the Linear discriminent and the Quadractic discriminent functions for Cases 1,2 and Case 3 respectively...